### PR TITLE
[CELEBORN-1530] support MPU for S3

### PR DESCRIPTION
**Pull request by zhaohehuhu**
### What changes were proposed in this pull request?

as title

### Why are the changes needed?
AWS S3 doesn't support append, so Celeborn had to copy the historical data from s3 to worker and write to s3 again, which heavily scales out the write. This PR implements a better solution via MPU to avoid copy-and-write.

### How was this patch tested?

WechatIMG257

I conducted an experiment with a 1GB input dataset to compare the performance of Celeborn using only S3 storage versus using SSD storage. The results showed that Celeborn with SSD storage was approximately three times faster than with only S3 storage.

Screenshot 2024-11-16 at 13 02 10

The above screenshot is the second test with 5000 mapper and reducer that I did.

### CODE REVIEW

**Comment by FMX** [comment]
Thanks for this PR. Are there any test results?

**Comment by zhaohehuhu** [comment]
> Thanks for this PR. Are there any test results?
Not yet. I'm ready to do a benchmark for it.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 1]
```diff
@@ -264,7 +299,9 @@ public void flush(boolean finalFlush, boolean fromEvict) throws IOException {
             } else if (diskFileInfo.isHdfs()) {
               task = new HdfsFlushTask(flushBuffer, diskFileInfo.getDfsPath(), notifier, true);
             } else if (diskFileInfo.isS3()) {
-              task = new S3FlushTask(flushBuffer, diskFileInfo.getDfsPath(), notifier, true);
+              task =
+                  new S3FlushTask(
+                      flushBuffer, notifier, true, s3MultipartUploadHandler, partNumber);
```
Comment (FMX) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:304]:
```suggestion
                      flushBuffer, notifier, true, s3MultipartUploadHandler, partNumber++);
```

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 2]
```diff
@@ -273,6 +310,7 @@ public void flush(boolean finalFlush, boolean fromEvict) throws IOException {
         if (task != null) {
           addTask(task);
           flushBuffer = null;
+          partNumber++;
```
Comment (FMX) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:313]: This line can be removed

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 3]
```diff
@@ -302,6 +340,11 @@ public void write(ByteBuf data) throws IOException {
       throw new AlreadyClosedException(msg);
     }
 
+    if (notifier.hasException() && s3MultipartUploadHandler != null) {
+      logger.warn("Abort s3 multipart upload for {}", diskFileInfo.getFilePath());
+      s3MultipartUploadHandler.complete();
+    }
+
     if (notifier.hasException()) {
```
Comment (FMX) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:348]: These two if blocks can be merged.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/MultipartUploadRequestParam.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.service.deploy.worker
+
+import java.lang.{Long => JLong}
+import java.util.{List => JList}
+
+case class MultipartUploadRequestParam(
```
Comment (FMX) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/MultipartUploadRequestParam.scala:23]: Unused class. Can be removed.

**Review by FMX — COMMENTED** [review]
Thanks for this PR but there are some points to polish.

**Comment by zyclove** [comment]
@zhaohehuhu @FMX @WillemJiang 
When can we expect this pr to be fully validated and merged into the main branch? It's very important.

**Comment by zhaohehuhu** [comment]
> @zhaohehuhu @FMX @WillemJiang When can we expect this pr to be fully validated and merged into the main branch? It's very important.

I still need more time to fully test it as S3 has some limitations related to MPU.

**Comment by FMX** [comment]
> @zhaohehuhu @FMX @WillemJiang When can we expect this pr to be fully validated and merged into the main branch? It's very important.

Every PR should be production-ready before it's been merged.

File: multipart-uploader/pom.xml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  <name>Celeborn Multipart Uploader</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-slf4j-impl</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-1.2-api</artifactId>
```
[the remark this answers is not in the document]
Comment (zhaohehuhu) [reply multipart-uploader/pom.xml]: done

File: multipart-uploader/pom.xml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+  <profiles>
+    <profile>
+      <id>aws-mpu</id>
+      <activation>
+        <property>
+          <name>aws-mpu-deps</name>
+        </property>
+      </activation>
+      <dependencies>
```
[the remark this answers is not in the document]
Comment (zhaohehuhu) [reply multipart-uploader/pom.xml]: done

File: multipart-uploader/pom.xml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-dependency-plugin</artifactId>
+      </plugin>
+    </plugins>
+  </build>
+
+  <profiles>
+    <profile>
+      <id>aws-mpu</id>
```
[the remark this answers is not in the document]
Comment (zhaohehuhu) [reply multipart-uploader/pom.xml]: done

File: service/src/main/java/org/apache/celeborn/server/common/service/mpu/bean/AWSCredentials.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.server.common.service.mpu.bean;
+
+public class AWSCredentials {
```
[the remark this answers is not in the document]
Comment (zhaohehuhu) [reply service/src/main/java/org/apache/celeborn/server/common/service/mpu/bean/AWSCredentials.java]: done

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 9]
```diff
@@ -187,6 +196,31 @@ public void initFileChannelsForDiskFile() throws IOException {
...
+          String key = diskFileInfo.getFilePath().substring(slashIndex + 1);
+
+          AWSCredentials awsCredentials =
+              new AWSCredentials(bucketName, s3AccessKey, s3SecretKey, s3EndpointRegion);
+          this.s3MultipartUploadHandler =
+              (MultipartUploadHandler)
+                  DynConstructors.builder()
+                      .impl(
+                          "org.apache.celeborn.S3MultipartUploadHandler",
+                          awsCredentials.getClass(),
```
[the remark this answers is not in the document]
Comment (zhaohehuhu) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java]: done

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java (hunk 1 above)
Comment (zhaohehuhu) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java]: done

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java (hunk 2 above)
Comment (zhaohehuhu) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java]: done

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java (hunk 3 above)
Comment (zhaohehuhu) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:362]: done

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/MultipartUploadRequestParam.scala (hunk 4 above)
Comment (zhaohehuhu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/MultipartUploadRequestParam.scala]: done

**Review by FMX — APPROVED** [review]
LGTM. Thanks. Merged into main(v0.6.0).

**Pull request closed without merging**